src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/Login/FeatureExtensions.cs

```diff
@@ -0,0 +1,513 @@
+
+            // write dataLen and options
+            await stream.TdsWriter.WriteIntAsync(dataLen, isAsync, ct).ConfigureAwait(false);
+            await stream.WriteByteAsync(options, isAsync, ct).ConfigureAwait(false);
+
+            // write accessToken for FedAuthLibrary.SecurityToken
+            switch (fedAuthFeatureData.libraryType)
+            {
+                case TdsEnums.FedAuthLibrary.MSAL:
+                    byte workflow = 0x00;
+                    switch (fedAuthFeatureData.authentication)
+                    {
+                        case SqlAuthenticationMethod.ActiveDirectoryPassword:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYPASSWORD;
+                            break;
+                        case SqlAuthenticationMethod.ActiveDirectoryIntegrated:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYINTEGRATED;
+                            break;
+                        case SqlAuthenticationMethod.ActiveDirectoryInteractive:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYINTERACTIVE;
+                            break;
+                        case SqlAuthenticationMethod.ActiveDirectoryServicePrincipal:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYSERVICEPRINCIPAL;
+                            break;
+                        case SqlAuthenticationMethod.ActiveDirectoryDeviceCodeFlow:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYDEVICECODEFLOW;
+                            break;
+                        case SqlAuthenticationMethod.ActiveDirectoryManagedIdentity:
+                        case SqlAuthenticationMethod.ActiveDirectoryMSI:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYMANAGEDIDENTITY;
+                            break;
+                        case SqlAuthenticationMethod.ActiveDirectoryDefault:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYDEFAULT;
+                            break;
+                        case SqlAuthenticationMethod.ActiveDirectoryWorkloadIdentity:
+                            workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYWORKLOADIDENTITY;
+                            break;
+                        default:
+                            if (context.AccessTokenCallback != null)
+                            {
+                                workflow = TdsEnums.MSALWORKFLOW_ACTIVEDIRECTORYTOKENCREDENTIAL;
+                            }
+                            else
+                            {
+                                Debug.Assert(false, "Unrecognized Authentication type for fedauth MSAL request");
+                            }
+                            break;
+                    }
+
+                    await stream.WriteByteAsync(workflow, isAsync, ct);
```

Comment:
You should only use TdsWriter to write IMO, if any API is needed, expose it directly from the TdsWriter.

Reply:
Changed all the instances I could find